packages/Search/src/details/DTK_DetailsDistributedSearchTreeImpl.hpp

```diff
@@ -197,99 +199,192 @@ DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
 template <typename DeviceType>
 template <typename Query>
 void DistributedSearchTreeImpl<DeviceType>::deviseStrategy(
-    Point eps, Kokkos::View<Query *, DeviceType> queries,
-    BVH<DeviceType> const &bvh, Kokkos::View<int *, DeviceType> &indices,
-    Kokkos::View<int *, DeviceType> &offset )
+    Kokkos::View<Query *, DeviceType> queries,
+    DistributedSearchTree<DeviceType> const &tree,
+    Kokkos::View<int *, DeviceType> &indices,
+    Kokkos::View<int *, DeviceType> &offset,
+    Kokkos::View<double *, DeviceType> & )
 {
-    int const n_queries = queries.extent( 0 );
-    Kokkos::View<Details::Overlap *, DeviceType> overlap_queries(
-        "overlap_queries", n_queries );
+    auto const &top_tree = tree._top_tree;
+    auto const &bottom_tree_sizes = tree._bottom_tree_sizes;
+
+    // Find the k nearest local trees.
+    top_tree.query( queries, indices, offset );
 
-    Kokkos::parallel_for( DTK_MARK_REGION( "fill_overlap_queries" ),
+    // Accumulate total leave count in the local trees until it reaches k which
+    // is the number of neighbors queried for.
+    auto const n_queries = queries.extent( 0 );
+    Kokkos::View<int *, DeviceType> _offset( offset.label(), n_queries + 1 );
+    Kokkos::deep_copy( _offset, 0 );
+    Kokkos::parallel_for(
+        DTK_MARK_REGION( "bottom_trees_with_required_cumulated_leaves_count" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n_queries ),
+        KOKKOS_LAMBDA( int i ) {
+            int leaves_count = 0;
+            for ( int j = offset( i ); j < offset( i + 1 ); ++j )
+            {
+                leaves_count += bottom_tree_sizes( indices( j ) );
+                ++_offset( i );
+                if ( leaves_count >= queries( i )._k )
```

Comment:
Is there anything we could/should do if there is not enough leaves?

Reply:
Now that I think about if we could also break if `bottom_tree_sizes( indices( j ) == 0 `.  No need to forward the query to empty trees.

Besides that no, not much you can do.  `query()` guarantees that if you have neighbors and it will find them.  The results that are returned are all valid entries, in other words, if you query for 10 neighbors and they are only 2 leaves, it will only returned the two indices sorted from the nearest to the farthest.

Reply:
What I was thinking but it's not clear from what I wrote, is what happens if I asked for 10 neighbors but you have only 8 points total? I guess it shouldn't happen in a real case but if someone has a toy problem they could create such a case.

Reply:
We are supposed to handle such cases.  I just looked at `tstDistributedSearchTree.cpp` and it might be a good occasion to add some test cases.

In short, say you only have one leave on rank 0, and that all other ranks have empty local trees.
Then if you query for 2 nearest neighbors (from any rank), you should expect offset={0,1}, indices={0}, ranks={0}

Reply:
ok thanks

Reply:
I added a couple tests.